lib/puppet/vendor/rgen_patch.rb

```diff
@@ -0,0 +1,129 @@
+# Patch to improve generated addXXX and setXXX method performance
+# Submitted to rgen as PR: https://github.com/mthiede/rgen/pull/22
+#
+# The current implementation of the generated addXXX method will iterate
+# over the contained array where each iteration will then test for object_id
+# equality. Once the iteration is done and such an entry is found it
+# checks it the added value is an instance of MMBase or MMGeneric. If it
+# is, then it's considered a duplicate and the method returns.
+#
+# This patch changes the implementation so that the test for MMBase/
+# MMGeneric is made prior to the iteration, thus avoiding the iteration
+# alltogether for all non model objects
+#
+# The patch also changes the setXXX method to allow it to optimize the way
+# it ensures that the added elements are unique. Calling the addXXX method
+# to do an increasingly expensive sequential scan is very expensive when
+# large sets of values are assigned.
+#
```

Comment:
worth mentioning and reference the PR to RGen and that this patch can then be removed.